libCacheSim/cache/eviction/ClockPro.c

```diff
@@ -0,0 +1,492 @@
+  }
+  if (params->hand_test == obj) {
+    params->hand_test = obj->queue.next;
+  }
+
+  cache_obj_t *hand_hot_next = params->hand_hot->queue.next;
+  move_obj_to_tail(&hand_hot_next, &params->hand_hot, obj);
+  params->hand_hot = obj->queue.next;
+
+  params->mem_hot -= obj->obj_size;
+  params->mem_cold += obj->obj_size;
+}
+
+static void ClockPro_promote(cache_t *cache, cache_obj_t *obj) {
+  ClockPro_params_t *params = (ClockPro_params_t *)cache->eviction_params;
+
+  if (obj->clockpro.status == TEST) {
+    if (params->mem_cold_max + (int64_t)obj->obj_size > cache->cache_size) {
+      params->mem_cold_max = cache->cache_size;
+    } else {
+      params->mem_cold_max += (int64_t)obj->obj_size;
+    }
+  }
+
+  while ((params->mem_hot + obj->obj_size) > (cache->cache_size - params->mem_cold_max)) {
+    ClockPro_run_hot(cache);
+  }
+
+  if (params->hand_cold == obj) {
+    params->hand_cold = obj->queue.next;
+  }
+  if (params->hand_test == obj) {
+    params->hand_test = obj->queue.next;
+  }
+
+  clockpro_status_e old_status = obj->clockpro.status;
+  obj->clockpro.status = HOT;
+  obj->clockpro.referenced = params->init_ref;
+  cache_obj_t *hand_hot_next = params->hand_hot->queue.next;
+  move_obj_to_tail(&hand_hot_next, &params->hand_hot, obj);
+  obj->queue.next  = hand_hot_next;
+  hand_hot_next->queue.prev = obj;
+
+  params->hand_hot = obj->queue.next;
+
+  if (old_status == COLD) {
+    params->mem_cold -= obj->obj_size;
+  } else if (old_status == TEST) {
+    params->mem_test -= obj->obj_size;
+  }
+
+  params->mem_hot += obj->obj_size;
+}
+
+// ***********************************************************************
+// ****                                                               ****
+// ****                  parameter set up functions                   ****
+// ****                                                               ****
+// ***********************************************************************
+static const char *ClockPro_current_params(cache_t *cache, ClockPro_params_t *params) {
+  static __thread char params_str[128];
+  snprintf(params_str, 128, "\n");
+  return params_str;
+}
+
+static void ClockPro_parse_params(cache_t *cache, const char *cache_specific_params) {
+  ClockPro_params_t *params = (ClockPro_params_t *)(cache->eviction_params);
+  char *params_str = strdup(cache_specific_params);
+  char *old_params_str = params_str;
+  char *end;
+
+  while (params_str != NULL && params_str[0] != '\0') {
+    char *key = strsep((char **)&params_str, "=");
+    char *value = strsep((char **)&params_str, ",");
+
+    while (params_str != NULL && *params_str == ' ') {
+      params_str++;
+    }
+
+    if (strcasecmp(key, "init-ref") == 0) {
+      params->init_ref = strtol(value, &end, 10);
+    } else if (strcasecmp(key, "init-ratio-cold") == 0) {
+      const double ratio = strtod(value, &end);
+      params->mem_cold_max = (int64_t)((double)cache->cache_size * ratio);
+    } else {
```

Comment:
we can also support `print` like [Clock.c](https://github.com/1a1a11a/libCacheSim/blob/9b407a57693d847616157f6279f7d0418da2397c/libCacheSim/cache/eviction/Clock.c#L319)